### PR TITLE
Adds error message with filename on Markdown error, fixes bug in panicOnBuild

### LINKED ISSUE #8524: Add more context to error reporting
## Summary

gatsby-transformer-remark returns errors without specifying in which file it happens.
It should display this information.

### Motivation

Context: I try to rewrite my website made with Hexo, using Gatsby.

One issue I stumbled upon is that my Markdown files were using a frontmatter with only three dashes at the end, working with Hexo, but ignored by Gatsby. So I had a mostly empty frontmatter in my GraphQL queries (only an empty title). I mention this in case it can help somebody with the same issue. Of course, it worked better once I added the three dashes at the start of the files...

My issue is that then, gatsby-transformer-remark reported Yaml errors. But it doesn't tell where they happened:

> > gatsby develop
> 
> success open and validate gatsby-config — 0.017 s
> success load plugins — 0.224 s
> success onPreInit — 0.844 s
> success delete html and css files from previous builds — 0.021 s
> success initialize cache — 0.023 s
> success copy gatsby files — 0.087 s
> success onPreBootstrap — 0.007 s
> error Plugin gatsby-transformer-remark returned an error
> 
> 
>   YAMLException: duplicated mapping key at line 18, column 2:
>        title: Atom
>        ^
> 
>   - loader.js:165 generateError
>     [PhiLhoSoft.GitHub.io]/[js-yaml]/lib/js-yaml/loader.js:165:10
> 
>   - loader.js:171 throwError
>     [PhiLhoSoft.GitHub.io]/[js-yaml]/lib/js-yaml/loader.js:171:9
> 
>   - loader.js:308 storeMappingPair
>     [PhiLhoSoft.GitHub.io]/[js-yaml]/lib/js-yaml/loader.js:308:7
> 
>   - loader.js:1071 readBlockMapping
>     [PhiLhoSoft.GitHub.io]/[js-yaml]/lib/js-yaml/loader.js:1071:9
> 
>   - loader.js:1332 composeNode
>     [PhiLhoSoft.GitHub.io]/[js-yaml]/lib/js-yaml/loader.js:1332:12
> 
>   - loader.js:1492 readDocument
>     [PhiLhoSoft.GitHub.io]/[js-yaml]/lib/js-yaml/loader.js:1492:3
> 
>   - loader.js:1548 loadDocuments
>     [PhiLhoSoft.GitHub.io]/[js-yaml]/lib/js-yaml/loader.js:1548:5
> 
>   - loader.js:1569 load
>     [PhiLhoSoft.GitHub.io]/[js-yaml]/lib/js-yaml/loader.js:1569:19
> 
>   - loader.js:1591 Object.safeLoad
>     [PhiLhoSoft.GitHub.io]/[js-yaml]/lib/js-yaml/loader.js:1591:10
> 
>   - parse.js:12 module.exports
>     [PhiLhoSoft.GitHub.io]/[gray-matter]/lib/parse.js:12:17
> 
>   - index.js:109 parseMatter
>     [PhiLhoSoft.GitHub.io]/[gray-matter]/index.js:109:17
> 
>   - index.js:50 matter
>     [PhiLhoSoft.GitHub.io]/[gray-matter]/index.js:50:10
> 
>   - on-node-create.js:31 Object.<anonymous>
>     [PhiLhoSoft.GitHub.io]/[gatsby-transformer-remark]/on-node-create.js:31:16

Here, the stack trace isn't very useful (but not a problem), but the "_YAMLException: duplicated mapping key at line 18, column 2:_" message just forgets to specify in which file this happens! Showing the path of the currently processed Markdown file would helps a lot!

Thanks for this very interesting tool. I struggle a bit to put bits and pieces in place, but I learn a lot in the process...

### PR DESCRIPTION
**Pull request by mslooten**
As discussed in #8524, when Markdown isn't correctly processed an error is thrown which doesn't include the filename. This fixes it using a try/catch with the `reporter.panicOnBuild` function, which is also changed since it had unexpected behavior (it didn't terminate the build, but it did terminate on other commands such as `gatsby develop`).

---edit by @pieh:
closes #8524

### CODE REVIEW

File: packages/gatsby-transformer-remark/src/on-node-create.js [hunk 1]
```diff
@@ -17,49 +17,56 @@ module.exports = async function onCreateNode(
...
+      .update(JSON.stringify(markdownNode))
+      .digest(`hex`)
 
-  createNode(markdownNode)
-  createParentChildLink({ parent: node, child: markdownNode })
+    createNode(markdownNode)
+    createParentChildLink({ parent: node, child: markdownNode })
+  } catch (err) {
+    reporter.panicOnBuild(
+      `Error processing Markdown file ${node.absolutePath}:\n
```
Comment (pieh) [inline packages/gatsby-transformer-remark/src/on-node-create.js]: I forgot about something earlier - markdown doesn't always come from file, so we will need to check if `node.absolutePath` is set before using it. For example:
```js
`Error processing Markdown ${node.absolutePath ? `file: ${node.absolutePath}` : ``}:\n`
```

**Comment by mslooten** [comment]
@pieh Thanks for pointing that out – hadn't considered that. Changed it up a bit so it checks if the path is available.

**Comment by ahinrichs** [comment]
What about giving some fallback identifier like node.id? This could help to find the error if the markdown isn't a file.

**Comment by pieh** [comment]
> What about giving some fallback identifier like node.id? This could help to find the error if the markdown isn't a file.

It definitely won't hurt. `id` will often be something like UUID (so not human readable), but I guess it's better feedback that none.

**Comment by mslooten** [comment]
Cool, can show `id` as fallback. Can I assume there's always an `id` available or should that also check for that presence?

**Comment by ahinrichs** [comment]
BTW, you could also on the fly start using the new `createContentDigest` API. See #8805.

**Comment by pieh** [comment]
> BTW, you could also on the fly start using the new `createContentDigest` API. See #8805.

Let's not derail this PR into adding more stuff here :)

> Cool, can show `id` as fallback. Can I assume there's always an `id` available or should that also check for that presence?

`id` is required, so it will always be there - it's safe to use that without any checks

**Comment by mslooten** [comment]
@pieh Great, thanks! Added the change to show the node.id.

**Pull request merged**